semstr/conversion/ptb.py

```diff
@@ -0,0 +1,282 @@
+
+    def __init__(self, token_id, value=None, lineno=None):
+        self.token_id = token_id
+        self.value = value
+        self.lineno = lineno
+
+    def __str__(self):
+        return "Token:'{tok}'{ln}".format(
+            tok=(self.value if self.value is not None else self._token_ids[self.token_id]),
+            ln=(':{}'.format(self.lineno) if self.lineno is not None else '')
+            )
+
+
+_token_pat = re.compile(r'\(|\)|[^()\s]+')
+def lex(line_or_lines):
+    """
+    Create a generator which returns tokens parsed from the input.
+    The input can be either a single string or a sequence of strings.
+    """
+
+    if isinstance(line_or_lines, str):
+        line_or_lines = [line_or_lines]
+
+    counter = 0
+    for n, line in enumerate(line_or_lines):
+        line.strip()
+        for m in _token_pat.finditer(line):
+            if m.group() == '(':
+                # if there is an extra '(' at the beginning of the sentence
+                if counter == 1:
+                    continue
+                counter += 1
+                yield Token(LPAREN_TOKEN)
+            elif m.group() == ')':
+                counter += 1
+                yield Token(RPAREN_TOKEN)
+            else:
+                counter += 1
+                yield Token(STRING_TOKEN, value=m.group())
+
+
+##################
+# Parser
+##################
+
+
+class Symbol:
+    _pat = re.compile(r'(?P<label>^[^0-9=-]+)|(?:-(?P<tag>[^0-9=-]+))|(?:=(?P<parind>[0-9]+))|(?:-(?P<coind>[0-9]+))')
+
+    def __init__(self, label):
+        self.label = label
+        self.tags = []
+        self.coindex = None
+        self.parindex = None
+        for m in self._pat.finditer(label):
+            if m.group('label'):
+                self.label = m.group('label')
+            elif m.group('tag'):
+                self.tags.append(m.group('tag'))
+            elif m.group('parind'):
+                self.parindex = m.group('parind')
+            elif m.group('coind'):
+                self.coindex = m.group('coind')
+
+    def simplify(self):
+        self.tags = []
+        self.coindex = None
+        self.parindex = None
+
+    def __str__(self):
+        return '{}{}{}{}'.format(
+            self.label,
+            ''.join('-{}'.format(t) for t in self.tags),
+            ('={}'.format(self.parindex) if self.parindex is not None else ''),
+            ('-{}'.format(self.coindex) if self.coindex is not None else '')
+        )
+
+
+PUNC = [',', '.']
+class Leaf:
+    def __init__(self, word, pos, parent_id, edge_tag = "Terminal", node_id = None):
+        self.word = word
+        self.pos = pos
+        self.parent_id = parent_id
+        self.edge_tag = edge_tag
+        self.tag = self.is_punc(word)
+        self.node_id = node_id
+        # self.is_punc = (self.pos == layer0.NodeTags.Punct)
+
+    def __str__(self):
+        return '({} {})'.format(self.pos, self.word)
+
+    def is_punc(self, word):
```

Comment:
This method's name is a little misleading, since it implies that it returns `True` or `False`. Maybe rename it to `get_node_tag` or something. Also, I think it's better to use the PTB tag to determine if a token is punctuation. See the list of tags: https://www.clips.uantwerpen.be/pages/mbsp-tags